File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxprod = nums[0], minprod = nums[0], ans = nums[0];
        int maxtemp, mintemp;

        for (int i = 1; i < nums.length; i++) {
            int n = nums[i];
            maxtemp = maxprod*n;
            mintemp = minprod*n;
            maxprod = Math.max(maxtemp, Math.max(mintemp, n));
            minprod = Math.min(maxtemp, Math.min(mintemp, n));
            ans = Math.max(ans, maxprod);
        }

        return ans;
    }
}

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        l = nums[0]
        s = nums[0]
        m = nums[0]
        for i in range(1, len(nums)):
            l_new = max(l * nums[i], s * nums[i], nums[i])
            s_new = min(l * nums[i], s * nums[i], nums[i])
            l = l_new
            s = s_new
            if l > m:
                m = l
        return m

        